hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeWithAdditionalFields.java

```diff
@@ -486,4 +498,237 @@ private int getGroupIDDB(String name) throws IOException {
+                return; // these should be intercepted in the xattr call
+              default: return;
+            } break;
+          default:
+            return;
+        }
+        //if we are here we are tracking
+        Optional<XAttr> xattr = Optional.empty();
+        logProvenanceEvent(parents.getL(), parents.getR(), op, xattr);
+      } break;
+      case PROJECT:
+      case OTHER:
+      default: return;
+    }
+  }
+  
+  @Override
+  public void logProvenanceEvent(FileProvenanceEntry.Operation op, XAttr xattr) throws IOException {
+    if(XAttr.NameSpace.PROVENANCE.equals(xattr.getNameSpace())) {
+      Pair<TrackedProv, INodeDirectory[]> parents = ProvenanceHelper.provenanceDirs(this);
+      switch(parents.getL()) {
+        case DATASET:
+        case HIVE:{
+          switch(parents.getR()[ProvParents.DATASET.ordinal()].getMetaStatus()) {
+            case FULL_PROV_ENABLED:
+            case MIN_PROV_ENABLED:
+              break;
+            default:
+              return;
+          }
+          //if we are here we are tracking
+          logProvenanceEvent(parents.getL(), parents.getR(), op, Optional.of(xattr));
+        } break;
+        case PROJECT:
+        case OTHER:
+        default: return;
+      }
+    }
+  }
+  
+  private void logProvenanceEvent(TrackedProv type, INodeDirectory[] parents, FileProvenanceEntry.Operation op,
+    Optional<XAttr> xattr) throws IOException {
+    
+    Long projectIId;
+    String projectName;
+    switch(type) {
+      case DATASET: {
+        projectIId = parents[ProvParents.PROJECT.ordinal()].getId();
+        projectName = parents[ProvParents.PROJECT.ordinal()].getLocalName();
+      } break;
+      case HIVE: {
+        projectIId = -1l;
+        projectName = ProvenanceHelper.projectNameFromHiveDB(
+          parents[ProvParents.DATASET.ordinal()].getLocalName());
+      } break;
+      default: return;
+    }
+    
+    UserGroupInformation ugi;
+    int remoteUserId;
+    String remoteUserName;
+    try {
+      ugi = NameNode.getRemoteUser();
+      remoteUserId = UsersGroups.getUserID(ugi.getUserName());
+      remoteUserName = ugi.getUserName();
+    } catch (IOException ex) {
+      String msg = "provenance - error getting user - issuer of operation for inode:" + this.getLocalName();
+      LOG.error(msg, ex);
+      throw ex;
+    }
+    String appId = ugi.getApplicationId();
+    if(appId == null) {
+      appId = "none";
+    }
+    
+    long timestamp = System.currentTimeMillis();
+    String p1Name = parents[ProvParents.PARENT_P1.ordinal()] != null
+      ? parents[ProvParents.PARENT_P1.ordinal()].getLocalName() : "";
+    String p2Name = parents[ProvParents.PARENT_P2.ordinal()] != null
+      ? parents[ProvParents.PARENT_P2.ordinal()].getLocalName() : "";
+    
+    String xattrName = xattr.isPresent() ? xattr.get().getName() : "";
+    
+    FileProvenanceEntry ple = new FileProvenanceEntry(id, op, logicalTime, timestamp, appId, remoteUserId,
```

Comment:
You don't increment the logical times at all? you are dependent on the log_metadata to increment it right?

Reply:
yes, I depend on the log_metatadata increment